README.md

```diff
@@ -152,69 +151,39 @@ Upon a call to `wasi_thread_spawn`, the WASI host must:
-    reserved values in wasi-libc.
-    For example, it can be used to indicate the main thread, which doesn't
-    have a TID in the current version of this proposal.
-
-### Process
-
-* A process is a group of threads.
-
-* The main thread starts with a process which only contains
-  the main thread.
-
-* Threads created by a thread in a process using `wasi_thread_spawn`
-  are added to the process.
-
-* When a thread is terminated, it's removed from the process.
-
-### Voluntary thread termination
-
-A thread can terminate itself voluntarily by returning from
-`wasi_thread_start`.
-
-### Changes to WASI `proc_exit`
-
-With this proposal, the `proc_exit` function takes extra responsibility
-to terminate all threads in the process, not only the calling one.
-
-Any of the threads in the process can call `proc_exit`.
-
-### Traps
-
-When a trap occurs in any thread, the entire process is terminated.
-
-### Process exit status
-
-If one or more threads call WASI `proc_exit` or raise a trap,
-one of them is chosen by the runtime to represent the exit status
-of the process.
-It's non deterministic which one is chosen.
-
-If all the threads in the process have been terminated without calling
-`proc_exit` or raising a trap, it's treated as if the last thread called
-`proc_exit` with exit code 0.
+#### Design choice: thread IDs
+
+When `wasi_thread_spawn` successfully spawns a thread, it returns a thread ID
+(TID) &mdash; 32-bit integer with several restrictions. TIDs are managed and
+provided by the WASI host. To avoid compatibility issues in wasi-libc, `0` is
+reserved; `wasi_thread_spawn` should not return it. To avoid leaking
+information, the host may choose to return random TIDs (as opposed to leaking OS
+TIDs).
+
+TIDs returned by `wasi_thread_spawn` must always zero the uppermost three bits:
+- the most significant bit is the sign bit; `wasi_thread_spawn` uses negative
+  values to indicate errors, so a valid TID must zero this bit (i.e., a valid
+  TID is always positive)
+- the second most significant bit is used for TID-based locking in wasi-libc
+  &mdash; this must be zeroed initially in a valid TID
+- the third most significant bit allows wasi-libc to mark threads as the main
+  thread; it is zeroed by any thread returned by `wasi_thread_spawn`
+
+#### Design choice: termination
+
+A `wasi-threads` module initially executes a single thread &mdash; the main
+thread. As `wasi_thread_spawn` is called, more threads begin to execute. Threads
+terminate in the following ways:
+
+- __voluntarily__, by returning from `wasi_thread_start`; other threads continue
+  to execute
+- __upon a trap__ in any thread; all threads are immediately terminated
+- __upon a `proc_exit` call__ in any thread; all threads are immediately
+  terminated.
+
+If multiple threads terminate simultaneously and for different reasons (e.g.,
+different `proc_exit` codes), one reason is chosen non-deterministically as the
+reason returned by the host.
```

Comment:
can you state that a return from `wasi_thread_start` is never chosen over other reasons?

Reply:
I would like to find a way to avoid even needing to state this.  One can think of calls to `proc_exit` and returning from a thread as acquiring some kind of global lock.  There can be race to acquire it but there is always going to be a winner, and so logically on one of them actually gets a chance to occur.

e.g. if multiple thread call proc_exit, its not so much that multiple proc_exit calls occur and host picks one.. its more then they race to be first caller and only one wins the race (like lock acquisition).   The thread that gets the lock exits the process.. none of the other proc_exit calls get a chance to be made.

I'm not sure how to best phase that in succinct way though.

Reply:
Its kind of redundant isn't it?   Its like saying "in the case of race, the first one wins" .. which is kind of the definition of what a race is. 

The way I see it once one thread calls proc_exit, all other threads are terminated so none of others get a chance to actually perform the proc_exit.

(I also don't see how you could possibly implement syscalls like `proc_exit` without some kind of lock.. but maybe I'm just not being imaginative enough :)

Reply:
Sure, verbiage suggestions then?

Reply:
I would maybe just drop this sentence, as I'm not sure exactly what including it is achieving.  But if you or @yamt feel like its useful to keep it for now thats fine with me.   We can always refine later.

Exactly how and when threads are shut down in response to `proc_exit` will probably need more fleshing out later anyway.

Reply:
Gladly! I agree with the "refine later" sentiment. It's too early to specify every detail when we don't yet know all the ways this can break. @yamt?

Reply:
honestly speaking, i don't understand why it is this controversial.
when i was writing the original text, i carefully tried to avoid describing implementation details.
i just described the only sane behavior i could think of.

Reply:
Ok, I'm going to drop it as I think it is the safer option here. What @sbc100 is getting at (I think) is that we should avoid over-specifying; the danger of that is that implementors of the spec get confused by too many restrictions and end up implementing something complex. I agree with @yamt that this is a sane way to specify things but I propose we wait to see where things go a bit more. E.g., I'm thinking about the difference between CLI usage and embedded use: in a CLI, I would choose to crash the entire process with `exit(<first observed proc_exit code>)` but in an embedding (think plugin) we don't want to crash the process, we actually want to stop all the threads (somehow) and return the code to the plugin caller.

Reply:
> E.g., I'm thinking about the difference between CLI usage and embedded use: in a CLI, I would choose to crash the entire process with `exit(<first observed proc_exit code>)` but in an embedding (think plugin) we don't want to crash the process, we actually want to stop all the threads (somehow) and return the code to the plugin caller. This second use case might end up requiring a very nuanced definition of termination.

FYI, my intention of the original text was to cover both types of implementations.
in toywasm, i implemented it in the latter way.